### DISCUSSION TITLE
Collaborate with other projects to create a new 3D interchange format

### DISCUSSION THREAD

**Discussion by emperorofmars** [godotengine/godot-proposals#11283] [3D]
**It is hard to get models out of Blender into Godot.**\
(correctly and without loosing half the work)

**I would like to propose for the Godot project to come together with other FOSS projects (and also non FOSS companies) to create a new extensible 3D interchange format.**

## Why

With version 4.4, Blender will have an animation system perfectly suitable for video games development. Unfortunately, exporting a model with plenty of animations, constraints, and what not, is not possible.

**FBX** is undocumented and not extensible.\
Blenders FBX exporter is faulty. It won't export keyframe tangents among many other issues.\
While Godot's new FBX importer works very well, I don't think this situation is at all good. It also doesn't support export.

Apart from FBX, what is there?
* **GLTF 2.0** is a format for runtime loading. I've written down in more detail why glTF 2.0 doesn't work for interchange here: https://gist.github.com/emperorofmars/d8abf0f4b9bd5434f9543511b243a254
* **USD** is not 'just a 3d format', and also not supported in Godot. Its scope is far bigger and more comparable to an entire Godot project rather than a single asset. It can make use of FBX files, and it could make use of this proposed format in the same way. USD does not replace FBX, or any other file format like that and, it's very complex.

## What should be done
Collaborate to create a file format that is highly **modular** and **extensible**.

It should:
* Be easy to implement.
* Represent 3d models **without loss** to the extent possible.
* Its **data-model** should match that of **authoring tools** like Blender, Godot, Bevy, Unreal Engine, Unity3d, Maya, 3dsMax, etc…
* Extensions should be easy to develop.
* Runtime loading efficiency should be a non concern.

This has also been proposed to:
* Blender: https://blender.community/c/rightclickselect/vZP5/?sorting=hot
* Bevy: https://github.com/bevyengine/bevy/discussions/16632

---

An ecosystem is always bigger than the sum of its parts.

This would enable a frictionless asset-pipeline between Blender and Godot, or really any other authoring tool.\
Supporting non FOSS tools would also be critically important in order for Godot to partake in the larger game development ecosystem.

Godot does this pretty well thanks to its FBX support, but it's now far easier to use 3dsMax or Maya than Blender with Godot.

### Proof of Concept
It consists of a container format with no specific features of its own.
All functionality is implemented as hot-loadable plugins, including core resource types like meshes.
A lot of effort was put into making plugins as self-contained as possible and easy to develop.

**Repository:** https://github.com/emperorofmars/stf-unity-poc \
**Explanation:** https://github.com/emperorofmars/stf-unity-poc/blob/master/STF/Docs/stf_format.md

### Arbitrary Material Definitions Proof of Concept
Current 3d formats support only some specific materials.
I've tried to create a material-definition format that supports arbitrary properties and is fully independent of any tool, game-engine or shader.

**Explanation & Repository:** https://github.com/emperorofmars/stf-unity-poc/blob/master/MTF/readme.md

### Extension System for FBX
This makes FBX (and other formats) extensible by serializing data into a 3d models object hierarchy. It's implemented for Blender and Unity.
It shows how easily an extension system can work on current existing formats.
This is at best a very heretical band-aid.

**Explanation & Repository:** https://github.com/emperorofmars/nna

**Comment by Zireael07** [comment]
Counter point: https://xkcd.com/927

**Reply by emperorofmars** [reply]
While I realize it's a moonshot, my argument is for projects to create one common format.
All alternatives are either dead or proprietary.

**Reply by Zireael07** [reply]
GLTF is not dead nor proprietary. USD is not dead either although not sure how open it is

**Reply by emperorofmars** [reply]
I account for why neither of these would work in the **'Why'** section of the original post.

OpenUSD is licensed under a very slightly modified Apache 2.0.

**Comment by AThousandShips** [comment]
Since there's already an open discussion this conversation should be continued there

**Reply by emperorofmars** [reply]
I genuinely searched before posting, and right now again. I am unable to find it, would you be so kind to link it, please?

All I could find was about internationalization, some specific existing format or whole game-engine projects. This is about 3d assets.

**Reply by AThousandShips** [reply]
My bad you linked one that looked like the one on here:
* #10045
